core/test/network-records-to-devtools-log.js

```diff
@@ -40,20 +46,181 @@ function headersArrayToHeadersDict(headersArray = []) {
+    endTime,
+    relativeTimingMax,
+  };
+}
+
+/**
+ * Takes the partial network timing and fills in the missing values so at least
+ * time is linear (if not realistic). The main timing properties on
+ * `NetworkRequest` and `requestTime` and `receiveHeadersEnd` on
+ * `NetworkRequest['timing']` will be computed. The other `timing` properties
+ * will not be automatically provided, but will be copied if provided, so they
+ * are up to the caller to order correctly. If no absolute times are provided,
+ * starts at `defaultStart`, finishes receiving headers `defaultTimingOffset` ms
+ * later, then endTime is `defaultTimingOffset` ms after that.
+ * Throws an error if conditions appear impossible to satisfy.
+ * @param {Partial<NetworkRequest>} values
+ * @return {NormalizedRequestTime}
+ */
+function getNormalizedRequestTiming(networkRecord) {
+  const extractedTimes = extractPartialTiming(networkRecord);
+
+  const possibleStarts = [
+    extractedTimes.startTime,
+    extractedTimes.requestTime,
+    extractedTimes.rendererStartTime,
+    // Note: since redirectResponseTimestamp (aka the redirect's endTime) is
+    // used as a last resort, some redirected requests may start before the
+    // redirect ends. Up to the caller to override if this matters.
+    extractedTimes.redirectResponseTimestamp,
+  ];
+  const startTime = possibleStarts.filter(timeDefined)[0] ?? defaultStart;
+
+  const rendererStartTime = timeDefined(extractedTimes.rendererStartTime) ?
+      extractedTimes.rendererStartTime :
+      // Because `rendererStartTime` was added much later, several tests assume
+      // requests start at `startTime`, so default to the same timestamp.
+      startTime;
+
+  // `startTime` and `requestTime` are the same.
+  const requestTime = startTime;
+
+  // `receiveHeadersEnd` is milliseconds after `requestTime`.
+  let {receiveHeadersEnd} = extractedTimes;
+  if (!timeDefined(receiveHeadersEnd)) {
+    if (timeDefined(extractedTimes.responseReceivedTime)) {
+      receiveHeadersEnd = extractedTimes.responseReceivedTime - requestTime;
+    } else if (timeDefined(extractedTimes.endTime)) {
+      // Pick a time between last defined `timing` (may just be `requestTime`) and `endTime`.
+      receiveHeadersEnd = (extractedTimes.relativeTimingMax +
+          (extractedTimes.endTime - requestTime)) / 2;
+    } else {
+      receiveHeadersEnd = Math.max(extractedTimes.relativeTimingMax, defaultTimingOffset);
+    }
+  }
+
+  const responseReceivedTime = timeDefined(extractedTimes.responseReceivedTime) ?
+      extractedTimes.responseReceivedTime : (requestTime + receiveHeadersEnd);
+
+  // endTime is allowed to be -1, e.g. for incomplete requests.
+  const endTime = extractedTimes.endTime ?? (responseReceivedTime + defaultTimingOffset);
+
+  return {
+    rendererStartTime,
+    startTime: startTime,
+    responseReceivedTime,
+    endTime,
+    timing: {
+      // TODO: other `timing` properties could have default values.
```

Comment:
this could be expanded, e.g. no DNS for follow up requests, no SSL for non-https, etc, but seemed unnecessary for now